File: native/Packages/HelixCore/Sources/HelixCore/Reports/Derived.swift
import Foundation

// ─────────────────────────────────────────────────────────────────────────────
// DERIVED METRICS — the arithmetic, kept behind a wall. A port of
// `src/lib/reports/derived.ts`.
//
// Every input is a figure already printed in the raw body; nothing is invented
// to fill a gap (a metric with no evidence is nil); it is pure and
// deterministic. It computes; it does not advise.
// ─────────────────────────────────────────────────────────────────────────────

public struct WeekDelta: Codable, Equatable, Sendable {
    public var label: String
    public var unit: String
    public var current: Double?
    public var previous: Double?
    /// Absolute change. Nil when either side is missing — not 0.
    public var delta: Double?
    /// Percentage change. Nil when the previous value is missing OR zero.
    public var pct: Double?
    public var digits: Int
    /// Print at FULL precision instead (tonnage).
    public var exact: Bool
}

public struct QualityTally: Codable, Equatable, Sendable {
    public var key: String
    public var label: String
    public var count: Int
}

public struct ExerciseProgression: Codable, Equatable, Sendable {
    public var name: String
    public var firstKg: Double
    public var lastKg: Double
    public var deltaKg: Double
    public var sessions: Int
}

public struct DerivedWeek: Codable, Equatable, Sendable {
    public var deltas: [WeekDelta]
    public var meanWorkingSetRpe: Double?
    public var ratedSets: Int
    public var workingSets: Int
    public var meanSetsPerSession: Double?
    public var meanVolumePerSessionKg: Double?
    public var failureSetShare: Double?
    public var quality: [QualityTally]
    public var flaggedSets: Int
    public var supplementsTaken: Double?
    public var supplementsPlanned: Double?
    public var fatigueReadings: Int
    public var fatigueSlots: Int
    public var domsDaysLogged: Int
    public var intakeDaysLogged: Int
    public var weighInDays: Int
    public var meanDeepPct: Double?
    public var meanRemPct: Double?
    public var meanAwakeMin: Double?
    public var progression: [ExerciseProgression]
    public var trainingDayKcal: Double?
    public var restDayKcal: Double?
}

/// Mean of the values that EXIST. Nil when none do — never 0.
func meanOf(_ xs: [Double?]) -> Double? {
    let ok = xs.compactMap { $0 }.filter(\.isFinite)
    return ok.isEmpty ? nil : ok.reduce(0, +) / Double(ok.count)
}

/// Sum of the values that exist. Nil when none do.
func sumOf(_ xs: [Double?]) -> Double? {
    let ok = xs.compactMap { $0 }.filter(\.isFinite)
    return ok.isEmpty ? nil : ok.reduce(0, +)
}

public enum Derived {
    /// The most recent EARLIER week in the ledger, anchored on the week's own start date.
    static func previousWeek(_ ledger: [LedgerWeek]?, weekStart: String) -> LedgerWeek? {
        guard let ledger, !ledger.isEmpty else { return nil }
        let earlier = ledger.filter { $0.weekStart < weekStart }
        guard let first = earlier.first else { return nil }
        return earlier.dropFirst().reduce(first) { $1.weekStart > $0.weekStart ? $1 : $0 }
    }

    static func deltaOf(_ label: String, _ unit: String, _ digits: Int, _ current: Double?, _ previous: Double?, exact: Bool = false) -> WeekDelta {
        let both = current?.isFinite == true && previous?.isFinite == true
        let c = current ?? 0, p = previous ?? 0
        return WeekDelta(
            label: label, unit: unit, current: current, previous: previous,
            delta: both ? c - p : nil,
            pct: both && p != 0 ? ((c - p) / abs(p)) * 100 : nil,
            digits: digits, exact: exact
        )
    }

    /// Top-set movement for the movements trained MORE THAN ONCE this week.
    static func progressionWithin(_ sessions: [ExportSession]) -> [ExerciseProgression] {
        var order: [String] = []
        var byName: [String: [(date: String, topKg: Double)]] = [:]
        for s in sessions {
            for e in s.exercises {
                let working = e.sets.filter { !$0.isWarmup && !$0.isGhost }
                guard let top = working.map(\.weightKg).max(), top.isFinite, top > 0 else { continue }
                if byName[e.name] == nil { order.append(e.name) }
                byName[e.name, default: []].append((s.date, top))
            }
        }
        var out: [ExerciseProgression] = []
        for name in order {
            let rows = byName[name]!
            if rows.count < 2 { continue }
            let sorted = rows.sorted { icuCompare($0.date, $1.date) < 0 }
            let firstKg = sorted[0].topKg, lastKg = sorted[sorted.count - 1].topKg
            out.append(ExerciseProgression(name: name, firstKg: firstKg, lastKg: lastKg, deltaKg: lastKg - firstKg, sessions: sorted.count))
        }
        return out.sorted { a, b in
            let da = abs(a.deltaKg), db = abs(b.deltaKg)
            if da != db { return da > db }
            return icuCompare(a.name, b.name) < 0
        }
    }

    static func intakeOn(_ days: [ExportDay], training: Bool) -> Double? {
        meanOf(days.filter { $0.isTrainingDay == training }.map(\.calories))
    }

    public static func week(_ input: WeeklyExportInput) -> DerivedWeek {
        let days = input.days
        let sessions = input.sessions
        let p = previousWeek(input.ledger, weekStart: input.weekStart)?.totals

        let totalVolume = sumOf(sessions.map(\.volumeKg))
        let deltas = [
            deltaOf("Total volume", "kg", 2, totalVolume, p?.totalVolumeKg, exact: true),
            deltaOf("Intake (avg/day)", "kcal", 0, meanOf(days.map(\.calories)), p?.avgKcal),
            deltaOf("Steps (avg/day)", "steps", 0, meanOf(days.map(\.steps)), p?.avgSteps),
            deltaOf("Bodyweight (avg)", "kg", 2, meanOf(days.map(\.weightKg)), p?.avgWeightKg),
            deltaOf("Water (avg/day)", "ml", 0, meanOf(days.map(\.waterMl)), p?.avgWaterMl),
            deltaOf("Cardio", "min", 0, sumOf((input.cardio ?? []).map(\.durationMin)), p?.cardioMinutes),
        ]

        var rpes: [Double] = []
        var workingSets = 0, failureSets = 0, flaggedSets = 0
        var qualityOrder: [String] = []
        var qualityCount: [String: Int] = [:]
        for s in sessions {
            for e in s.exercises {
                for set in e.sets {
                    if set.isWarmup || set.isGhost { continue }
                    workingSets += 1
                    if let r = set.rpe, r.isFinite { rpes.append(r) }
                    if set.failure { failureSets += 1 }
                    if let q = set.quality, !q.isEmpty {
                        flaggedSets += 1
                        if qualityCount[q] == nil { qualityOrder.append(q) }
                        qualityCount[q, default: 0] += 1
                    }
                }
            }
        }
        let quality = qualityOrder
            .map { QualityTally(key: $0, label: SetTags.quality[$0]?.label ?? $0, count: qualityCount[$0]!) }
            .sorted { a, b in a.count != b.count ? a.count > b.count : icuCompare(a.key, b.key) < 0 }

        var deepPcts: [Double] = [], remPcts: [Double] = []
        for d in days {
            guard let sleep = d.sleepMin, sleep > 0 else { continue }
            if let deep = d.deepMin { deepPcts.append((deep / sleep) * 100) }
            if let rem = d.remMin { remPcts.append((rem / sleep) * 100) }
        }

        return DerivedWeek(
            deltas: deltas,
            meanWorkingSetRpe: meanOf(rpes),
            ratedSets: rpes.count,
            workingSets: workingSets,
            meanSetsPerSession: sessions.isEmpty ? nil : Double(workingSets) / Double(sessions.count),
            meanVolumePerSessionKg: !sessions.isEmpty && totalVolume != nil ? totalVolume! / Double(sessions.count) : nil,
            failureSetShare: workingSets > 0 ? (Double(failureSets) / Double(workingSets)) * 100 : nil,
            quality: quality,
            flaggedSets: flaggedSets,
            supplementsTaken: sumOf(days.map(\.supplementsTaken)),
            supplementsPlanned: sumOf(days.map(\.supplementsPlanned)),
            fatigueReadings: (input.fatigue ?? []).count,
            fatigueSlots: days.count * 3,
            domsDaysLogged: Set(input.doms.map(\.date)).count,
            intakeDaysLogged: days.filter { $0.calories != nil }.count,
            weighInDays: days.filter { $0.weightKg != nil }.count,
            meanDeepPct: meanOf(deepPcts),
            meanRemPct: meanOf(remPcts),
            meanAwakeMin: meanOf(days.map(\.awakeMin)),
            progression: progressionWithin(sessions),
            trainingDayKcal: intakeOn(days, training: true),
            restDayKcal: intakeOn(days, training: false)
        )
    }
}

/// `String.prototype.localeCompare` as the web's ICU root collation orders the
/// strings this domain sorts: whitespace before punctuation before digits
/// before letters, letters compared case-insensitively first and lower-case
/// first on a tie. An approximation of ICU, sufficient for exercise names,
/// clock times and snake_case keys; anything exotic would need real ICU.
func icuCompare(_ a: String, _ b: String) -> Int {
    func cls(_ c: Character) -> (Int, String) {
        if c.isWhitespace { return (0, String(c)) }
        if c.isNumber { return (2, String(c)) }
        if c.isLetter { return (3, c.lowercased()) }
        return (1, String(c))
    }
    let ka = a.map(cls), kb = b.map(cls)
    for (x, y) in zip(ka, kb) where x != y {
        if x.0 != y.0 { return x.0 < y.0 ? -1 : 1 }
        return x.1 < y.1 ? -1 : 1
    }
    if ka.count != kb.count { return ka.count < kb.count ? -1 : 1 }
    // Tertiary: same letters, different case — lower-case first.
    for (x, y) in zip(a, b) where x != y {
        return x.isLowercase ? -1 : 1
    }
    return 0
}

File: src/lib/reports/derived.ts
/**
 * DERIVED METRICS — the arithmetic, kept behind a wall.
 *
 * ── WHY THIS IS A SEPARATE MODULE AND A SEPARATE SECTION ─────────────────────
 * `weeklyExport.ts` opens with a rule it has kept for its whole life: every
 * number in the document is one the app MEASURED, and nothing is "derived,
 * averaged into existence, or estimated to fill a column". That rule is the
 * reason the export is trustworthy, and it is not up for renegotiation.
 *
 * But a reader — a coach, a model, a spreadsheet — wants deltas, rates and
 * coverage, and computing them from the raw body every time is work that will
 * be done inconsistently or not at all. So the answer is not to relax the rule;
 * it is to give the derived figures their own fenced-off address, arriving
 * AFTER every measurement they are built from, under a heading that says out
 * loud what they are.
 *
 * Three properties hold for everything in here:
 *
 *  · Every input is a figure already printed in the raw body above. Nothing in
 *    this file reaches for a table the export did not already show the reader.
 *  · Nothing is invented to fill a gap. A metric with no evidence returns null
 *    and its line says so, exactly as the raw body does with `—`.
 *  · It is pure and deterministic — same payload, same numbers, no clock.
 *
 * The one thing this file must never become is a coach. It computes; it does
 * not advise. "Tonnage +8.9%" is arithmetic. "Tonnage is up, keep going" is an
 * opinion, and opinions belong to whoever reads this, not to the exporter.
 */
import type { WeeklyExportInput, ExportDay, ExportFatigue, ExportSession, LedgerWeek } from '@/lib/reports/weeklyExport'
import { SET_QUALITY } from '@/lib/training/setTags'
import { computeMorningCharge, sleepQualityParts, stressParts } from '@/lib/scoring/battery'
import { FATIGUE_SLOTS, SLOT_LABEL, fatigueLevel } from '@/lib/recovery/fatigue'

/** Mean of the values that EXIST. Null when none do — never 0. */
function mean(xs: Array<number | null | undefined>): number | null {
  const ok = xs.filter((v): v is number => v != null && Number.isFinite(v))
  return ok.length ? ok.reduce((a, b) => a + b, 0) / ok.length : null
}

/** Sum of the values that exist. Null when none do. */
function total(xs: Array<number | null | undefined>): number | null {
  const ok = xs.filter((v): v is number => v != null && Number.isFinite(v))
  return ok.length ? ok.reduce((a, b) => a + b, 0) : null
}

/** One week-over-week comparison, with the change stated both ways. */
export interface WeekDelta {
  label: string
  unit: string
  current: number | null
  previous: number | null
  /** Absolute change. Null when either side is missing — not 0. */
  delta: number | null
  /** Percentage change. Null when the previous value is missing OR zero. */
  pct: number | null
  /** Decimal places the figure is worth printing at. */
  digits: number
  /**
   * Print at FULL precision instead, trailing zeros dropped.
   *
   * Tonnage only. It is a sum of quarter-kilogram microloads, so "26340.00" is
   * two digits of noise and "8329.25" is two digits of real work — the same
   * reason `exact()` exists in the renderer, and the same reason a fixed
   * decimal count cannot serve both.
   */
  exact?: boolean
}

/** How many working sets carried each technique flag. */
export interface QualityTally {
  key: string
  label: string
  count: number
}

/** One movement's top-set movement WITHIN this week. */
export interface ExerciseProgression {
  name: string
  firstKg: number
  lastKg: number
  deltaKg: number
  sessions: number
}

/**
 * One day's battery v8 inputs, exactly as the scorer read them.
 *
 * The app shows `appPct`; this is everything behind it. The four `q` terms and
 * the three stress terms are the scorer's own functions run on the day line's
 * figures plus the two baselines the payload carries for the purpose — so a
 * reader can see WHICH input moved the number, which the number cannot say.
 */
export interface BatteryDay {
  date: string
  weekdayLabel: string
  /** `daily_scores.battery_pct`. Null when the app never scored the day. */
  appPct: number | null
  morningCharge: number
  ratio: number
  stagesQ: number
  hrvQ: number
  rhrQ: number
  onsetTrouble: boolean
  stress: number
  rhrTerm: number
  hrvTerm: number
  fatigueTerm: number
  /** The latest fatigue reading's word, or null when none was logged. */
  fatigueLabel: string | null
}

export interface DerivedWeek {
  deltas: WeekDelta[]
  battery: BatteryDay[]
  /** Mean effort across every RATED working set — not the session-level rating. */
  meanWorkingSetRpe: number | null
  ratedSets: number
  workingSets: number
  meanSetsPerSession: number | null
  meanVolumePerSessionKg: number | null
  failureSetShare: number | null
  quality: QualityTally[]
  flaggedSets: number
  supplementsTaken: number | null
  supplementsPlanned: number | null
  fatigueReadings: number
  fatigueSlots: number
  domsDaysLogged: number
  intakeDaysLogged: number
  weighInDays: number
  meanDeepPct: number | null
  meanRemPct: number | null
  meanAwakeMin: number | null
  progression: ExerciseProgression[]
  trainingDayKcal: number | null
  restDayKcal: number | null
}

/**
 * The previous week's totals, from the ledger.
 *
 * The ledger is EVERY week of the programme, oldest first, and the exported
 * week is normally its last row — but not always: a re-export of an older week
 * would leave later rows sitting after it. So the comparison is anchored on the
 * week's own start date rather than on array position, and returns null when
 * this week is the first the programme has (where "no previous week" is the
 * honest answer and a zero would be a fabricated one).
 */
function previousWeek(ledger: readonly LedgerWeek[] | undefined, weekStart: string): LedgerWeek | null {
  if (!ledger?.length) return null
  const earlier = ledger.filter((w) => w.weekStart < weekStart)
  if (!earlier.length) return null
  return earlier.reduce((best, w) => (w.weekStart > best.weekStart ? w : best))
}

function deltaOf(
  label: string, unit: string, digits: number,
  current: number | null, previous: number | null,
  exact = false,
): WeekDelta {
  const both = current != null && Number.isFinite(current)
    && previous != null && Number.isFinite(previous)
  return {
    label, unit, digits, exact, current, previous,
    delta: both ? current - previous : null,
    // A percentage against zero is a division by zero, not an infinite
    // improvement — the raw delta beside it already tells that story.
    pct: both && previous !== 0 ? ((current - previous) / Math.abs(previous)) * 100 : null,
  }
}

/**
 * Top-set movement for the movements trained MORE THAN ONCE this week.
 *
 * Deliberately narrow. A real progression reading compares an exercise to the
 * last time it was performed, which may be nine days ago and outside this
 * payload — and reaching for it would mean this module fetching data the raw
 * body never showed, which is the one thing it must not do. So the claim is
 * scoped to what the week can prove: a movement trained twice in seven days,
 * and what its heaviest working set did between the two.
 *
 * Warm-ups and ghosts are excluded — neither defines a top load.
 */
function progressionWithin(sessions: readonly ExportSession[]): ExerciseProgression[] {
  const byName = new Map<string, Array<{ date: string; topKg: number }>>()
  for (const s of sessions) {
    for (const e of s.exercises) {
      const working = e.sets.filter((set) => !set.warmup && !set.ghost)
      const top = working.length ? Math.max(...working.map((set) => set.weightKg)) : null
      // An unloaded movement has no top LOAD to move; its progression lives in
      // reps, which is a different axis and not one this summary claims.
      if (top == null || !Number.isFinite(top) || top <= 0) continue
      const bucket = byName.get(e.name) ?? []
      bucket.push({ date: s.date, topKg: top })
      byName.set(e.name, bucket)
    }
  }
  const out: ExerciseProgression[] = []
  for (const [name, rows] of byName) {
    if (rows.length < 2) continue
    const sorted = [...rows].sort((a, b) => a.date.localeCompare(b.date))
    const firstKg = sorted[0].topKg
    const lastKg = sorted[sorted.length - 1].topKg
    out.push({ name, firstKg, lastKg, deltaKg: lastKg - firstKg, sessions: sorted.length })
  }
  // Biggest absolute movement first; ties broken by name so the order is stable.
  return out.sort((a, b) => Math.abs(b.deltaKg) - Math.abs(a.deltaKg) || a.name.localeCompare(b.name))
}

/**
 * The battery's view of each day. The LATEST fatigue slot wins, as
 * `latestFatigue` decides it — slots are compared by their position in the
 * day, not by their label's alphabet.
 */
function batteryDays(input: WeeklyExportInput): BatteryDay[] {
  const slotIndex = (label: string) => FATIGUE_SLOTS.findIndex((s) => SLOT_LABEL[s] === label)
  return input.days.map((d) => {
    const latest = (input.fatigue ?? [])
      .filter((f) => f.date === d.date)
      .reduce<ExportFatigue | null>((best, f) => (best == null || slotIndex(f.slot) > slotIndex(best.slot) ? f : best), null)
    const signals = {
      sleepHours: (d.sleepMin ?? 0) / 60, deepMinutes: d.deepMin ?? 0, remMinutes: d.remMin ?? 0,
      sleepGoalHours: input.sleepGoalHours ?? 8,
      restingHR: d.restingHr ?? undefined, baselineHR: d.restingHrBaseline ?? undefined,
      hrvMs: d.hrvMs ?? undefined, hrvBaseline: d.hrvBaseline ?? undefined,
      fatigueLevel: latest?.level ?? null,
    }
    const q = sleepQualityParts(signals)
    const st = stressParts(signals)
    const onsetTrouble = d.sleepOnsetTrouble === true
    return {
      date: d.date, weekdayLabel: d.weekdayLabel,
      appPct: d.batteryPct ?? null,
      morningCharge: computeMorningCharge(q.quality, onsetTrouble),
      ratio: q.ratio, stagesQ: q.stagesQ, hrvQ: q.hrvQ, rhrQ: q.rhrQ, onsetTrouble,
      stress: st.drain, rhrTerm: st.rhrTerm, hrvTerm: st.hrvTerm, fatigueTerm: st.fatigueTerm,
      fatigueLabel: latest ? (fatigueLevel(latest.level)?.label ?? String(latest.level)) : null,
    }
  })
}

/** Mean intake across the days matching a training/rest predicate. */
function intakeOn(days: readonly ExportDay[], training: boolean): number | null {
  return mean(days.filter((d) => d.isTrainingDay === training).map((d) => d.calories))
}

export function derivedWeek(input: WeeklyExportInput): DerivedWeek {
  const days = input.days
  const sessions = input.sessions
  const prev = previousWeek(input.ledger, input.weekStart)
  const p = prev?.totals

  const totalVolume = total(sessions.map((s) => s.volumeKg))
  const deltas: WeekDelta[] = [
    deltaOf('Total volume', 'kg', 2, totalVolume, p?.totalVolumeKg ?? null, true),
    deltaOf('Intake (avg/day)', 'kcal', 0, mean(days.map((d) => d.calories)), p?.avgKcal ?? null),
    deltaOf('Steps (avg/day)', 'steps', 0, mean(days.map((d) => d.steps)), p?.avgSteps ?? null),
    deltaOf('Bodyweight (avg)', 'kg', 2, mean(days.map((d) => d.weightKg)), p?.avgWeightKg ?? null),
    deltaOf('Water (avg/day)', 'ml', 0, mean(days.map((d) => d.waterMl)), p?.avgWaterMl ?? null),
    deltaOf('Cardio', 'min', 0, total((input.cardio ?? []).map((c) => c.durationMin)), p?.cardioMinutes ?? null),
  ]

  // ── Set-level tallies, in ONE pass ──
  const rpes: number[] = []
  let workingSets = 0, failureSets = 0, flaggedSets = 0
  const qualityCount = new Map<string, number>()
  for (const s of sessions) {
    for (const e of s.exercises) {
      for (const set of e.sets) {
        // Same exclusion the RPE-coverage figure uses upstream: a warm-up is
        // never rated by design and a ghost did not happen.
        if (set.warmup || set.ghost) continue
        workingSets += 1
        if (set.rpe != null && Number.isFinite(set.rpe)) rpes.push(set.rpe)
        if (set.failure) failureSets += 1
        if (set.quality) {
          flaggedSets += 1
          qualityCount.set(set.quality, (qualityCount.get(set.quality) ?? 0) + 1)
        }
      }
    }
  }

  const quality: QualityTally[] = [...qualityCount.entries()]
    .map(([key, count]) => ({ key, label: SET_QUALITY[key]?.label ?? key, count }))
    .sort((a, b) => b.count - a.count || a.key.localeCompare(b.key))

  // ── Sleep architecture as SHARES, not minutes ──
  // 39 minutes of deep sleep means one thing after 9h and another after 5h30,
  // and the raw body already prints the minutes. The share is the reading the
  // minutes cannot give on their own.
  const deepPcts: number[] = [], remPcts: number[] = []
  for (const d of days) {
    if (d.sleepMin == null || !(d.sleepMin > 0)) continue
    if (d.deepMin != null) deepPcts.push((d.deepMin / d.sleepMin) * 100)
    if (d.remMin != null) remPcts.push((d.remMin / d.sleepMin) * 100)
  }

  const supplementsTaken = total(days.map((d) => d.supplementsTaken))
  const supplementsPlanned = total(days.map((d) => d.supplementsPlanned))

  return {
    deltas,
    battery: batteryDays(input),
    meanWorkingSetRpe: mean(rpes),
    ratedSets: rpes.length,
    workingSets,
    meanSetsPerSession: sessions.length ? workingSets / sessions.length : null,
    meanVolumePerSessionKg: sessions.length && totalVolume != null ? totalVolume / sessions.length : null,
    failureSetShare: workingSets ? (failureSets / workingSets) * 100 : null,
    quality,
    flaggedSets,
    supplementsTaken,
    supplementsPlanned,
    fatigueReadings: (input.fatigue ?? []).length,
    // Seven days, four slots. The denominator is fixed rather than counted, so
    // a week with no readings at all still reports 0 of 28 instead of 0 of 0.
    // Three slots a day, whichever kind of day it is — a training day asks
    // Waking / Before / After and a rest day asks Waking / Midday / Night, so
    // the denominator does not depend on the week's shape.
    fatigueSlots: days.length * 3,
    domsDaysLogged: new Set(input.doms.map((d) => d.date)).size,
    intakeDaysLogged: days.filter((d) => d.calories != null).length,
    weighInDays: days.filter((d) => d.weightKg != null).length,
    meanDeepPct: mean(deepPcts),
    meanRemPct: mean(remPcts),
    meanAwakeMin: mean(days.map((d) => d.awakeMin)),
    progression: progressionWithin(sessions),
    trainingDayKcal: intakeOn(days, true),
    restDayKcal: intakeOn(days, false),
  }
}
